types.go

```diff
@@ -0,0 +1,296 @@
+	// Healthy is whether or not the server is healthy according to the current
+	// Autopilot config.
+	Healthy bool
+
+	// StableSince is the last time this server's Healthy value changed.
+	StableSince time.Time
+}
+
+// IsStable returns true if the ServerState shows a stable, passing state
+// according to the given AutopilotConfig
+func (h *ServerHealth) IsStable(now time.Time, minStableDuration time.Duration) bool {
+	if h == nil {
+		return false
+	}
+
+	if !h.Healthy {
+		return false
+	}
+
+	if now.Sub(h.StableSince) < minStableDuration {
+		return false
+	}
+
+	return true
+}
+
+// ServerStats holds miscellaneous Raft metrics for a server
+type ServerStats struct {
+	// LastContact is the time since this node's last contact with the leader.
+	LastContact time.Duration
+
+	// LastTerm is the highest leader term this server has a record of in its Raft log.
+	LastTerm uint64
+
+	// LastIndex is the last log index this server has a record of in its Raft log.
+	LastIndex uint64
+}
+
+type State struct {
+	startTime        time.Time
+	Healthy          bool
+	FailureTolerance int
+	Servers          map[raft.ServerID]*ServerState
+	Leader           raft.ServerID
+	Voters           []raft.ServerID
+	Ext              interface{}
+}
+
+func (s *State) ServerStabilizationTime(c *Config) time.Duration {
+	// Only use the configured stabilization time when autopilot has
+	// been running for 110% of the configured stabilization time.
+	// Before that time we haven't been running long enough to
+	// be able to take these values into account. 110% is pretty
+	// arbitrary but with the default config would prevent the
+	// stabilization time from mattering for an extra second. This
+	// allows for leeway in how quickly we get the healthy RPC responses
+	// after autopilot is started.
+	if time.Since(s.startTime) > (c.ServerStabilizationTime*110)/100 {
+		return c.ServerStabilizationTime
+	}
+
+	// ignore stabilization time if autopilot hasn't been running long enough
+	// to be tracking any server long enough to meet that requirement
+	return 0
+}
+
+// Raft is the interface of all the methods on the Raft type that autopilot needs to function. Autopilot will
+// take in an interface for Raft instead of a concrete type to allow for dependency injection in tests.
+type Raft interface {
+	AddNonvoter(id raft.ServerID, address raft.ServerAddress, prevIndex uint64, timeout time.Duration) raft.IndexFuture
+	AddVoter(id raft.ServerID, address raft.ServerAddress, prevIndex uint64, timeout time.Duration) raft.IndexFuture
+	DemoteVoter(id raft.ServerID, prevIndex uint64, timeout time.Duration) raft.IndexFuture
+	LastIndex() uint64
+	Leader() raft.ServerAddress
+	GetConfiguration() raft.ConfigurationFuture
+	RemoveServer(id raft.ServerID, prevIndex uint64, timeout time.Duration) raft.IndexFuture
+	Stats() map[string]string
+	LeadershipTransferToServer(id raft.ServerID, address raft.ServerAddress) raft.Future
+}
+
+type Delegate interface {
```

Comment:
I wonder if we can find a term for this interface that is slightly more descriptive. 

The last 3 methods are about Servers, and Config and Notify are pretty generic concepts. Maybe this interface could say something about servers? `ServerStore` or maybe even `Servers` ?

Reply:
Really this is application integration. There are some bits like how the autopilot config is stored or how we determine what servers exist that are left up to the application to provide at runtime.

We could call it `Servers` but IMO that is not really more descriptive about why the interface exists and how it will be used.

We could call it `ApplicationIntegration` which at least makes it clear that these are the bits the application needs to provide for autopilot to do its job.

Reply:
I like the word `Application` here, to help indicate the need to hooks/callbacks into the supervisory process. Whether `ApplicationIntegration`/`ApplicationDelegate`/`Application`, I don't have a strong opinion.